ramips: add support for TP-Link Archer C2

franzflasch opened the pull request:
It was made by Serge Vasilugin <vasilugin@yandex.ru> based on these patches:
https://patchwork.ozlabs.org/patch/894715/
https://pastebin.com/UEjhWJap

Specification:
- System-On-Chip:	MT7620A
- CPU/Speed: 580 MHz
- Flash-Chip: Winbond 25Q64BVSIG
- Flash size: 8192 KiB
- RAM: 64 MiB
- Wireless No1: SoC-integrated: MT7620A 2.4GHz 802.11bgn
- Wireless No2: On-board chip: MT7610EN 5GHz 802.11ac
- Switch: RTL8367RB Gigabit Switch
- USB: Yes 1 x 2.0

Installation Instructions can be found here:
https://openwrt.org/toh/tp-link/ac750#installation_v1_model_only

I've successfully tested it with my Archer C2 V1.1.

I also want to mention that with this patch only the 2.4GHz wifi works. 5GHz can only be used with some non-free SW from mediatek.

Signed-off-by: Franz Flasch <franz.flasch@gmx.at>

robimarko commented:
Add specs and install instructions to the commit

franzflasch commented:
OK, done. Thanks for the hint

orangepizza commented:
It's too late to this get into openwrt 18.06, right?
and I moved the openwrt forum thread to lede forum.
https://forum.lede-project.org/t/support-for-tp-link-archer-c2-v1/14301

mkresin commented:
Please move the rtl8366 changes into an extra commit with a commit message explaining what the changes are supposed to do.

mkresin commented on target/linux/ramips/dts/ArcherC2.dts:
```diff
@@ -0,0 +1,168 @@
+/include/ "mt7620a.dtsi"
+
+/ {
+	compatible = "ralink,mt7620a-soc";
```
should be `compatible = "tplink,c2", "ralink,mt7620a-soc";`

mkresin commented on target/linux/ramips/base-files/lib/ramips.sh:
```diff
@@ -85,6 +85,9 @@ ramips_board_detect() {
 	*"C108")
 		name="c108"
 		;;
+	*"C2")
```
drop this and use `tplink,c2` as userspace boardname (from the dts compatible) everywhere.

franzflasch replied:
Sorry but I don't understand how this should work. Please explain in more detail how this should work

john-tho replied:
In these other files, `tplink,c2` is wanted, not just `c2`:
- target/linux/ramips/base-files/etc/board.d/01_leds
- target/linux/ramips/base-files/etc/board.d/02_network
- target/linux/ramips/base-files/etc/diag.sh
- target/linux/ramips/base-files/lib/upgrade/platform.sh

You may need to move your code to fix ordering.

Example https://github.com/openwrt/openwrt/commit/5bdd238b634a34b038fbea9046893de29a77c000

mkresin commented on target/linux/ramips/base-files/etc/board.d/01_leds:
```diff
@@ -94,6 +94,12 @@ c108)
 	ucidef_set_led_netdev "modem" "modem" "$boardname:green:modem" "wwan0"
 	;;
+c2)
+	ucidef_set_led_switch "lan" "lan" "$board:green:lan" "switch1" "0x1e"
```
$boardname please, similar to all other LEDs in this file.

mkresin commented on target/linux/ramips/dts/ArcherC2.dts:
```diff
@@ -0,0 +1,168 @@
+			reg = <0x20000 0x7a0000>;
+		};
+
+		partition@7c0000 {
```
all partitions beginning from here should be read-only. I don't see a reason to change the contents of the partitions.

mkresin commented on target/linux/ramips/dts/ArcherC2.dts:
```diff
@@ -0,0 +1,168 @@
+/dts-v1/;
+
+/include/ "mt7620a.dtsi"
+
```
Please include `dt-bindings/gpio/gpio.h` and `dt-bindings/input/input.h` as all other dts files are doing.

Use the GPIO_ACTIVE_LOW and GPIO_ACTIVE_HIGH macros afterwards in stead of 1 and 0 in the gpio parameters.

Same for the [key code macros](https://elixir.bootlin.com/linux/v4.14.41/source/include/uapi/linux/input-event-codes.h#L572) instead of the hex values.

Check the recent [ramips board additions](https://github.com/lede-project/source/commits/master/target/linux/ramips) for examples.

mkresin commented on target/linux/ramips/dts/ArcherC2.dts:
```diff
@@ -0,0 +1,168 @@
+		mt76@0,0 {
+			reg = <0x0000 0 0 0 0>;
+			device_type = "pci";
+			mediatek,mtd-eeprom = <&radio 32768>;
```
please use a hex value for the eeprom offset as well.

mkresin commented on target/linux/ramips/dts/ArcherC2.dts:
```diff
@@ -0,0 +1,168 @@
+	pinctrl {
+		state_default: pinctrl0 {
+			gpio {
+				ralink,group = "i2c", "uartf", "wled", "nd_sd", "ephy", "spi refclk";
```
drop the `nd_sd`. None of the nd_sd pins are used as GPIO.

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -45,6 +45,19 @@ define Device/Archer
   IMAGE/sysupgrade.bin := tplink-v2-image -s -e | append-metadata
 endef
 
+define Device/ArcherC2
```
Should be `define Device/tplink_c2`

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -45,6 +45,19 @@ define Device/Archer
+  DEVICE_TITLE := TP-Link ArcherC2
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport kmod-switch-rtl8366-smi kmod-switch-rtl8367b
+endef
+TARGET_DEVICES += ArcherC2
```
Should be `TARGET_DEVICES += tplink_c2`

mkresin commented on target/linux/ramips/image/mt7620.mk:
```diff
@@ -45,6 +45,19 @@ define Device/Archer
+define Device/ArcherC2
+  $(Device/Archer)
+  DTS := ArcherC2
+  SUPPORTED_DEVICES := c2
```
can be dropped if the correct Device define is used

mkresin reviewed (changes requested):
Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

Please be prepared for more change requests, your code misses quite a few changes which were applied to all dts files during the last year and I'm not sure if I spotted already everything.

mkresin commented:
Please update the commit title. It should be something like `ramips: add support for TP-Link Archer C2`. 

The installation instructions should be in the commit message and not a link to some external resources.

franzflasch committed:
kernel: rtl8366-smi: add Realtek switch management via mii-bus

Current version of rtl8366-smi module only supports Realtek switch
managment via two gpio lines. This adds Realtek switch
management via mii_bus. Tested on a Tp-link Archer C2 v1 (Mediatek
SoC mt7620a based)

dts-file configuration should look like this:

rtl8367rb {
	compatible = "realtek,rtl8367b", "rtl8367b";
	realtek,extif1 = <1 0 1 1 1 1 1 1 2>;
	mii-bus = <&mdio0>;
};

&ethernet {
	status = "okay";
	mtd-mac-address = <&rom 0xf100>;
	pinctrl-names = "default";
	pinctrl-0 = <&rgmii1_pins &rgmii2_pins &mdio_pins>;

	port@5 {
		status = "okay";
		mediatek,fixed-link = <1000 1 1 1>;
		phy-mode = "rgmii";
	};

	mdio0: mdio-bus {
		status = "okay";
	};
};

Original author of this changeset is Serge Vasilugin <vasilugin@yandex.ru>

Signed-off-by: Franz Flasch <franz.flasch@gmx.at>

franzflasch committed:
ramips: add support for TP-Link Archer C2

Specification:

- System-On-Chip: MT7620A
- CPU/Speed: 580 MHz
- Flash-Chip: Winbond 25Q64BVSIG
- Flash size: 8192 KiB
- RAM: 64 MiB
- Wireless No1: SoC-integrated: MT7620A 2.4GHz 802.11bgn
- Wireless No2: On-board chip: MT7610EN 5GHz 802.11ac
- Switch: RTL8367RB Gigabit Switch
- USB: Yes 1 x 2.0

Preparing a TFTP recovery image for initial flashing:

Currently the only method to install openwrt for the first time is via
TFTP download in u-boot. After first install you can use regular updates.
WARNING: This method also overwrites the bootloader partition!

Create a TFTP recovery image:

1) Download a stock TP-Link Firmware file here:
   https://www.tp-link.com/en/download/Archer-C2_V1.html#Firmware
2) Extract it and remove 512 byte header:
   #> dd if=c2v1_stock_firmware.bin of=ArcherC2V1_tp_recovery.bin bs=1 skip=512
3) Extract u-boot from the binary file:
   #> dd if=c2v1_stock_firmware.bin of=c2v1_uboot.bin bs=1 skip=512 count=131072
4) Now merge the sysupgrade image and the u-boot into one binary:
   #> cat c2v1_uboot.bin openwrt-squashfs-sysupgrade.bin > ArcherC2V1_tp_recovery.bin

The resulting image can be flashed via TFTP recovery mode.

There is also an old prepared openwrt tftp image available here:
http://beinder.at/pub/c2_recovery.zip

Flash instructions:

1) To flash the recovery image, start a TFTP server from IP address 192.168.0.66 and
   serve the recovery image named ArcherC2V1_tp_recovery.bin.
2) Connect your device to the LAN port, then press the WPS/Reset button and power it up.
   Keep pressing the WPS/Reset button for 10 seconds.
   It will try to download the recovery image and flash it.

It can take up to 20-25 minutes to finish. When it reaches 100%, the router will reboot itself.

Original author of this changeset is Serge Vasilugin vasilugin@yandex.ru

Signed-off-by: Franz Flasch <franz.flasch@gmx.at>

franzflasch commented:
> Please move the rtl8366 changes into an extra commit with a commit message explaining what the changes are supposed to do.

OK, I've now prepared 2 commits one for the rtl8366 changes ans one for the Archer C2 support, which depends on these changes.

Currently I've not incorporated your comments, will do this today or tomorrow.

franzflasch commented:
OK, @mkresin I've now incorporated your comments. There are 2 things I do not understand on your comments, would be nice if you could explain it a bit more in detail.

john-tho commented:
Hey @franzflasch
You need to amend your commit messages, and fix the first line (with a following blank link) of the commit (the commit title).

Commit message for the first commit (rtl8366 modifications) should start with:
kernel: rtl8366-smi: add Realtek switch management via mii-bus

Second commit should start with:
ramips: add support for TP-Link Archer C2

You should probably also use
git commit --amend --author='Serge Vasilugin <vasilugin@yandex.ru>'
for the correct attribution?

I have been using these patches with my C2, and would like to see them integrated.

Is it worth bumping the email thread, or emailing Serge Vasilugin, to get the changes requested in the email thread clarified or implemented?

franzflasch commented:
Hey @john-tho 

Thanks for the hints, I've added all your suggestions. I just do not understand how board detection now works without the entry in the file "target/linux/ramips/base-files/lib/ramips.sh"

Is it parsed from the device tree? Would be nice if someone could explain this.

I've removed the c2 entry from target/linux/ramips/base-files/lib/ramips.sh anyways and implemented it like it was done in the example commit.

franzflasch commented:
@john-tho Could you also test the current changes? Just to be sure the current changes do not brake anything. I am currently not able to test those changes on my Archer C2.

john-tho commented:
Hey @franzflasch,

Thanks,
I will give it a try.

> board detection
> parsed from the device tree

This looks like it: ramips specific https://github.com/openwrt/openwrt/commit/3eb839b5795e6f8d6461f94dbc4217c3641e7f19. removed: https://github.com/openwrt/openwrt/commit/7d3ec5f8332859be4c5da6275df1cbdfad6e73, for generic:
https://github.com/openwrt/openwrt/blob/master/package/base-files/files/lib/preinit/02_sysinfo

https://github.com/openwrt/openwrt/blob/78f1974bc565d7544589a49ad8efd92c4ddec5b3/package/base-files/files/lib/preinit/02_sysinfo#L5

There are a few lines in the ArcherC2.dts file with trailing whitespace.

john-tho commented:
This seemed to work fine for me,
I used the patch from this pull request and compiled OpenWrt 18.06-SNAPSHOT, r6907+79-7e15e21766
Did sysupgrade -n, and the switch and 2.5G wifi are working.

franzflasch commented:
@john-tho OK, great! Hopefully someone merges it asap.

lars18th commented:
Hi,

Please, take note that support for **TP-Link TL-WR842N v5** is included for the next release 18.06:
http://github.com/openwrt/openwrt/commit/5881e2434fce53724c5d09b365945429a4985c41
But the first PR is pushed only 4 days ago:
http://github.com/openwrt/openwrt/pull/1045

So, if the developers of these patches will fix the _small_ pending changes (after more than a month), then the support for the Acher C2 will too be included in time.

From my tests, these patches are working in my Archer C2 v1.
Regards.

franzflasch commented:
Hi @lars18th !

How can I compile the recovery image? I couldn't find any information on that.

Regards.

orangepizza commented:
@franzflasch 
>Hi lars18th ! 
>How can I compile the recovery image? I couldn't find any information on that. 
>Regards.

assume it has same profile as c20i, (I think it's same?) 

> 
> TFTP recovery images for Archer C20i differ from the normal firmware images as they do not have the 512-byte long header at the beginning of the file. In order to create a recovery image, download a stock firmware image like this one and strip the header from the file:
> 
> dd if=archer_firwmare.bin of=ArcherC20iV1_tp_recovery.bin bs=512 skip=1

but need somebody with spi clip to safely test it.

lars18th commented:
@franzflasch,

You only need to make a more "generic" firmware build!

The _guide_ to compile the firmware for this device is described in this case in `mt76x8.mk`. The problem is that your patch doesn't describe how to generate the files `factory.bin` & `tftp-recovery.bin`. And this is because you defined it as `Device/tplink_c2`. Instead of this is preferable to use the generic profile `Device/tplink` (see at the top of the file: http://github.com/heinzek/openwrt/blob/master/target/linux/ramips/image/mt76x8.mk)

You inly need to adapt your patch to this "template".

See the example for the Similar Archer C50 v3 (at file `target/linux/ramips/image/mt76x8.mk`):
http://github.com/openwrt/openwrt/pull/704/files

franzflasch commented:
Hi @lars18th !

My patch generates the factory image just fine. The only thing missing is the recovery image. The link you provided above doesn't describe how it is generated. 

The command only strips the header of the original TP Link recovery image, but does not generate a working openwrt recovery image.

However there is another link on this page which describes how it may work:
https://pwassi.privatedns.org/lede/archerc20i/

Will try and test this. Problem is, that I don't have much time at the moment...

Regards,
Franz

mkresin commented:
Some more remarks. There are four different versions of the C2. The PR seem to target v1. Hence v1 should be in devicetree compatible string and in the userspace boardname.

> You only need to make a more "generic" firmware build!

Be careful with such an advice! I was confused regarding the installation instructions as well, poked a few people and got the following reply:

The tp-link tftp recovery is a mess and more like a trap.

The C20i for example writes an image received via tftp right to the start of the flash. Regardless whether a bootloader is prepended or not. If you use a "normal" factory.bin, the bootloader would be overwritten with the actual image. If you use a 0x00 padded tftp-recovery.bin, the bootloader is overwritten with the 0x00. I somehow suspect we are in a similar situation here and the mentioned recovery.bin has a bootloader prepended. More informations can be found in https://forum.openwrt.org/viewtopic.php?id=60414&p=1#p296894.

Of course, there are way more implementations of the tftp recovery function. For some boards the bootloader doesn't get overwritten if the to be written data are only 0x00 (tftp-recovery.bin). Other boards don't touch the bootloader at all and are fine with the "normal" factory.bin.

Flashing OpenWrt - using the TP-Link webinterface + factory.bin - might not work as well. Usually the TP-Link webinterface doesn't touch the bootloader partition if no bootloader is bundled with the image (yet). Nevertheless, it expects/enforces a bundled bootloader for some boards. The webinterface upgrade writes some bytes to a fixed position of the image, which normally is right after the bootloader. In case of the openwrt factory.bin, we don't have a bootloader bundled, but the kernel at the position. Hence some bytes of the kernel are changed, which cause a boot error/crash on boot.

To find a working flash method, the advice to have a spi clip + a backup of the flash chip is essential. Nevertheless, someone need to find a acceptable installation method. Relying on an external image from a random source - which might disappear at some point in time - is something at least I don't like.

lars18th commented:
Hi @franzflasch and @mkresin and @orangepizza ,

In order to clarify and summarize some relevant questions:

1. Number of HW revisions: We target (at time) only **v1**.
2. The TFTP method for _first_ flash (after that you can use the `sysupgrade.bin` for a new flash)... it **really overwrites the bootloader**.
3. For TFTP flashing some "ancient" (but working) versions exist. For example you can use the `openwrt_r50020\ArcherC2V1_tp_recovery.bin` inside "c2_recovery.zip" (get it from: http://github.com/orangepizza/openwrt-archer-c2-lede17.01/releases).
4. To create a working `ArcherC2V1_tp_recovery.bin` you can use the guide from @orangepizza .

And if you need a stock mtd0 / u-boot partition dump, then I attach one from my stock device:
[archerC2v1-bootloader.zip](https://github.com/openwrt/openwrt/files/2115909/archerC2v1-bootloader.zip)

Regards.

mkresin commented:
> The TFTP method for first flash (after that you can use the sysupgrade.bin for a new flash)... it really overwrites the bootloader.

Here we go. This is the bare minimum I expect it to be mentioned in the commit message. I mean, till now I had no clue why an external file is required though we're building an factory.bin (which obviously doesn't work).

lars18th commented:
Hi @mkresin ,

Yes, I know it! And for this reason I recommend to take _three_ actions to start to merge the support for this device:

1. Merge the patch about the switch management via the mii-bus (rtl8366-smi). This change is transparent to other devices. Futhermore, in the future others can benefit from it if the hardware is routed in this way. I don't see any reason to not merge it now.
2. Modify the HEADER (commit message) to indicate with a **crystal clear** description that the TFTP method **overwrites the bootloader**. As an workaround, I accept to temporally include a link for the "c2_recovery.zip" file until finish action 3 (next).
3. Include support to generate the `ArcherC2V1_tp_recovery.bin` file using the `sysupgrade.bin` and the `bootloader.image`. I feel @orangepizza can help on this, as he has done it in the past (he is the creator of the "c2_recovery.zip" file).

Please, commet if you agree with this roadmap.

franzflasch commented:
OK, thanks for pointing this out. 

I think the best solution in this case would be to remove the factory image completely from the build and find out how the recovery images is generated. This, in my point of view would be the 
cleanest solution.

mkresin commented:
The easiest would be to attach to the serial console and write the sysupgrade image right from the u-boot prompt. The following should work:

```
tftpboot 0x80a00000 openwrt-sysupgrade.bin
erase 0xbc020000 +0x7a0000
cp.b 0x80a00000 0xb0020000 0x7a0000
```

It loads an image via tftp to memory address 0x80a00000. 0x7a0000 bytes of the flash, starting from offset 0x20000, is erased. It should exactly match the firmware partition. Finally, the data (0x7a0000 bytes) stored at memory address 0x80a00000 is written to offset 0x20000 of the flash chip.

It should work in all cases and be mentioned in the commit message, as it is the only reliable (first) installation method for now. The recovery image approach (+ external file) can be mentioned in the wiki. An OpenWrt build recovery image can be added later.

lars18th commented:
Hi @mkresin ,

> The easiest would be to attach to the serial console and write the sysupgrade image right from the u-boot prompt...

Not really!
Perheps this isn't the best location to comment it, however I do it:

- Flashing from the bootloader console over the serial using the TFTP server, it's the last resolt for flashing the firmware (using the device). But is not the best solution. The bootloader many times isn't sufficient reliable for this task. A more intelligent solution is to boot a firmware from memory downloaded via a TFTP server. Inside this firmware you can write the flash more reliable. However, this has two drawbacks: first the firmware needs to have the option for boot from RAM, that's the **initramfs** version, that by default isn't available for download. The second problem is with non-writable partitions. In this case you need to load the "mtd-rw" kernel module (with `insmod mtd-rw.ko i_want_a_brick=1`). Then you can flash the new image without troubles. And more important... prior to flash you can to a backup dump and copy it to a safe place.

That's my five cents. :smile:

orangepizza commented:
to @dbeinder
if you get this message, can you tell us how you made the recovery image of archer C2 (v1)?

@lars18th 
I don't know how to make such recovery image, that c2_recovery.zip is made by @yoq in forum.openwrt.org, (@dbeinder on github) witch is no longer works.  he runs beinder.at, but it doesn't help as I needed to pass recaptur v1

dbeinder commented:
@orangepizza 
I haven't touched openwrt in over two years, but this is what I have from back then:
 - The C2v1 TFTP recovery overwrites the uboot partition, hence it must be included in a recovery.bin
 - An official stock TPLink firmware file has a 512 byte header, which needs to be stripped for TFTP:
`dd if=c2v1_stock_firmware.bin of=ArcherC2V1_tp_recovery.bin bs=1 skip=512`
 - To create a recovery for openwrt, simply stick the stock uboot in front of the squashfs:
`dd if=c2v1_stock_firmware.bin of=c2v1_uboot.bin bs=1 skip=512 count=131072`
`cat c2v1_uboot.bin openwrt-squashfs-sysupgrade.bin > ArcherC2V1_tp_recovery.bin`
not sure about the elegance of dropping a 128k binary into the openwrt build system though...

lars18th commented:
Hi @dbeinder ,

Thank you for your explanation about how to generate a `tftp-recovery.bin` image for the Archer C2 (v1). :+1: 

However, for @orangepizza , I need to explain some questions:

- Inside OpenWRT is a "regular" tplink template _with_ a firmware generation tool called `mktplinkfw2.c`.
- See the description of the _generic_ templete about how to create the `tftp-recovery.bin` image here: https://github.com/openwrt/openwrt/blob/14cb05909a8dc6e42008697b4fe47b8fa00789d8/target/linux/ramips/image/mt76x8.mk#L15
- The `mktplinkfw2.c` tool is here: https://github.com/openwrt/openwrt/blob/master/tools/firmware-utils/src/mktplinkfw2.c
- For older devices a version 1 of the tool exists too: https://github.com/openwrt/openwrt/blob/master/tools/firmware-utils/src/mktplinkfw.c

So, these are my _assumptions_:

- If you use the tool `mktplinkfw2.c` with correct parameters, then you can generate a `factory.bin` image compatible with the stock firmware.
- If you use the `pad-extra` command with **512k** then you can create a `tftp-recovery.bin` image compatible with the TFTP recovery method.
- The previous sentence implies that with a correct header the flash will not replace the bootloader, and instead of it will skip this part. I comment this because I don't see inside the OpenWRT source any "uboot" dump for TP-Link devices. We need to check this first!

In any case, all of this is only for generating a `factory.bin` and `tftp-recovery.bin` images. That's an advanced improvement. The first step is merge the "rtl8366-smi" patch, and the initial support. I hope @mkresin will accept it, and he merge first the "rtl8366-smi" part; and @franzflasch will rewrite the description of the initial support with a description about how to create a `tftp-recovery.bin` and a temporal (but reliable) URL to download one valid version... until the "mktplinkfw2" solution is committed.

You agree?
Remember: _We're short in time to add this device for the 18.06 version!_

orangepizza commented:
@lars18th 
I didn't make that recovery image. @dbeinder did. 
Pad-extra command won't help us here. it will just zero-fill the uboot partition with zeros, bricking the device. and he extracted uboot part of stock image`
 dd if=c2v1_stock_firmware.bin of=ArcherC2V1_tp_recovery.bin bs=1 skip=512`
then put the sysupgrade firmware behind it.
`cat c2v1_uboot.bin openwrt-squashfs-sysupgrade.bin > ArcherC2V1_tp_recovery.bin`

unless we upload the uboot partition to repository, it can't make tftp recovery firmware.

he's first line
`dd if=c2v1_stock_firmware.bin of=ArcherC2V1_tp_recovery.bin bs=1 skip=512`
is for make back-to-stock tftp recovery image from stock firmware files(because just renameing it for tftp will brick the device for the god's sake!)

franzflasch commented:
Hi!

Sorry, but I still don't get it. How is this TFTP feature of TP-link even supposed to work? 

Why do I need to strip the 512 byte header in order to use the TFTP feature? This can not be the intention of TP-Link as I assume that you can also use it to recover the device from a broken firmware upgrade. Or do they have special recover images for TFTP?

lars18th commented:
Hi @orangepizza ,

> I didn't make that recovery image. @dbeinder did.

I know it! He commented here. :smile: 

> Pad-extra command won't help us here...

I doesn't help _if_ the current bootloader code will replace _all the time_ the bootloader. My **assumption** is that the code skips these zeros when the header indicates to "not override the uboot section". Why I think this? Because other TP-Link devices have the same TFTP recovery method for flashing a firmware. And none of them has a `tftp-recovery.bin` image created from a uboot dump. Perhaps I'm wrong, but I feel the in the header that puts the "mktplinkfw2" tool is for this reason.

@franzflasch ,
> Sorry, but I still don't get it. How is this TFTP feature of TP-link even supposed to work?

The TFTP feature is a "last resolt method" for flashing a firmware. It runs from the bootloader. And at time, until we can't generate a `factory.bin` firmware with a correct header to flash using the stock firmware it's the only method for flashing the OpenWRT firmware without opening the device.

> Why do I need to strip the 512 byte header in order to use the TFTP feature? 

This represents the header that uses the stock firmware to recognize a valid firmware. The TFTP method uses a "dump" firmware, but the STOCK method checks the 512 byte header. But, please take note that this is only true for _STOCK_ firmwares.

In any case: Why not merge first the support and work after on this _advanced_ functions?

franzflasch commented:
> The TFTP method uses a "dump" firmware, but the STOCK method checks the 512 byte header.

OK, understood. Is the TFTP method then only intended for internal use and not for the customer?

> In any case: Why not merge first the support and work after on this advanced functions?

OK, sounds good, I will update the git comment with a link of dbeinders image and a description on how to generate such TFTP capable recovery image.

orangepizza commented:
I'll check if stock firmware gui does accept openwrt sysupgrade. But wouldn't stock firmware have header means it will look for thoses header? Maybe old version don't?

dbeinder commented:
Yes, i don't think a tftp image(recovery) is possible without putting the bootloader into the openwrt source. But a factory.bin (to install openwrt via web from stock) should work. Check out how its done for the Archer C20i, which has almost the same hardware, and has the same problem with tftp:
https://github.com/openwrt/openwrt/blob/openwrt-18.06/target/linux/ramips/image/mt7620.mk#L48

franzflasch commented:
@orangepizza 

> well best way to support is complie uboot for this router by ourselves. but that's entirly different beast.

u-boot source code can be downloaded at tplink: https://www.tp-link.com/de/download/Archer-C2_V1.html#GPL-Code

So should not be too difficult to build it ourselves. However I am not sure if it really works, I once built the u-boot from the 'official' sources for a netgear router which didn't work. I think they don't maintain the downloadable GPL source code very well.

dbeinder commented:
In any case, the boot loader is not supposed to be part of openwrt anyway, so i would do the same as with C20i: don't build the tftp-recovery image as part of the build process and leave instructions on how to do it manually in the wiki.
But you can generate factory images (might be neccessary to have separate for EU and US, see C50) so there is a way for a first install. The factory header is created by mktplinkfw2 (called by tplink-v2-image).

mkresin commented:
> Sorry, but I still don't get it. How is this TFTP feature of TP-link even supposed to work?

It isn't. According to TP-Link Poland, the recovery mode is not for user, it's used only in factory.

>> The easiest would be to attach to the serial console and write the sysupgrade image right from the u-boot prompt...

> Not really!

It is for sure not the most convenient way. But testing my instructions and adding them to the commit message is the easiest way to get the PR merged. I expect that quite some time is required to find another working method. The installation via bootloader command line is usually the last effort but it always works.

I already outline that an upgrade via the tp-link webinterface using the factory image might not work. We were bitten by the issue last October around the time of the OpenWrt Summit.

> The first step is merge the "rtl8366-smi" patch, and the initial support. I hope @mkresin will accept it, and he merge first the "rtl8366-smi" part

I'm not going to commit parts of the PR. Albeit the changes look good to me, I'm concerned that they will introduce a regression for other boads using the rtl8366 driver. For this reason these changes will be first commited to master, to be tested for a few weeks.

> Remember: We're short in time to add this device for the 18.06 version!

See above. Quite unlikely that the PR will made it into the release branch.

lars18th commented:
Hi @mkresin ,

> I'm not going to commit parts of the PR. Albeit the changes look good to me, I'm concerned that they will introduce a regression for other boads using the rtl8366 driver. For this reason these changes will be first commited to master, to be tested for a few weeks.

I agree with the merge to "master". Almost this will be an advance until we can complete the support. :smile: 

Only to summarize:
- The "rtl8366-smi" patch is completed. We feel that it doesn't introduce a regression with other devices, as the only one using this connection is this one (until now). However it will be tested in the "master" branch for a minimum of two weeks.
- The patch for the support for the Archer C2 (v1) is _almost_ completed. The only required action is to update the flashing guide.

It's this right?

mkresin commented:
> - The "rtl8366-smi" patch is completed. We feel that it doesn't introduce a regression with other devices, as the only one using this connection is this one (until now). However it will be tested in the "master" branch for a minimum of two weeks.
> - The patch for the support for the Archer C2 (v1) is _almost_ completed. The only required action is to update the flashing guide.

The summary is correct. The flashing instructions need to be tested and the commit message updated accordingly.

franzflasch commented:
OK, I've now updated the instructions. Hopefully everyone is happy now.

I didn't test it though, I don't have the possibilty to test it right now. So if someone could manage to test it, that would be great.

Regards.

lars18th commented:
Thank you @franzflasch . Good Job!

I only suggest a minimal change for better understanding:

```
Preparing a TFTP recovery image for initial flashing:

Currently the only method to install openwrt for the first time is via 
TFTP download in u-boot. After first install you can use regular updates.
WARNING: This method also overwrites the bootloader partition!
```

I hope you agree!

franzflasch commented:
@lars18th 

OK, done. I really appreciate all the help, thanks!

lars18th commented:
Hi @mkresin ,

So, prepared to merge with "master"?

dbeinder commented:
I'd suggest renaming the device to `c2-v1` since there is already a v3 on the shelfs with completely different hardware.
Also, step 2) is independent and just creates a back to stock tftp file and is not neccessary to create a openwrt recovery file. Step 3) includes skip=512 and extracts the uboot partition on its own.

lars18th commented:
> I'd suggest renaming the device to c2-v1 since there is already a v3 on the shelfs with completely different hardware.

Good idea! Please, @franzflasch !

mkresin commented:
I merged a modified version of the PR to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary). It already includes the change of the boardname from c2 to c2-v1.

Furthermore the sysupgrade was broken, as userspace boardname and SUPPORTED_DEVICES wasn't in sync.

There were some issues in the switch driver patch which I fixed. Beginning from logical errors like missing brackets for the change in rtl8366_smi_init(), code style warnings discovered by ./scripts/checkpatch.pl to compiler warnings.

I moved the change of the kng_rc.dts into an own commit with a proper explanation.

Please review the changes in my staging carefully! Not unlikely that I introduced a new issue. Please runtime test an image build from staging tree and report the results.

franzflasch commented:
@mkresin You are right, there may are some cosmetical issues, however I don't see any logical error, can you please show where the logical error was? As you've commited into your own repo using gitweb it is very inconvinient to make a comparison.

Anyways, let's bring the changes into master asap.

mkresin commented:
> however I don't see any logical error

Great, I misread the part of the code due to bad indentation. There isn't an error and the {} I added aren't required. But at least it explains why nobody reported the compile error I would have expected. I updated my staging accordingly.

>  Anyways, let's bring the changes into master asap.

I'm not going to commit anything untested. I asked for the review of my code + the runtime test of an image build from my staging tree for a reason.

orangepizza commented:
it throws sysupgrade error from old (unofficial) versions. But it's expected, since boardname changed and it's not necessary to honer unoffical builds. 
after TFTP flash, it works normally

Gruntruck commented:
Hi guys. I have a couple of C2 v1's I use as AP's and here are my findings: TFTP method is unreliable. Two of my devices flashed quickly with TFTP. Third took multiple retries (at 10kbyte/sec) and last one only accepted data at 500 bytes/sec, never succeeding with flash despite multiple re-tries. I used all tricks in the book (lowering Ethernet speed, using different sizes of anticipation window etc) :( So if this is to work for Joe-sixpack then factory update is a must.

Anyway, I bought a flash programmer with SPI clip and copied/flashed Winbond 8MB Flash directly on the motherboard. So if you want me to check something that risks bricking the router, I can do it for you.

I tried orangepizza's 18.06 snapshot: (https://expirebox.com/download/3aaedd3688623fcfa8172da3836e9791.html ) and while it worked, 2.4 was still buggy and 5GHz didn't work.

Tell me if you need something new tested...

lars18th commented:
Hi @Gruntruck ,

Thank you for helping! :smile: 

Regarding your comments about troubles with the TFTP flash, I **repeat** my opinion about it:
- Flashing the Flash chip is only _reliable_ using a running OpenWRT or stock firmware.
- The TFTP flash needs to be a "last resolt" when no other option is avaiable.
- I _strongly_ recommend to generate `initramfs` images for booting using the bootloader console and RAM. This way uses the same TFTP loading but doesn't flash the chip, so it's safe. Futhermore, from the same RAM booted image it's possible to write another firmware using a reliable method.

And regarding questions around this device:

> I tried orangepizza's 18.06 snapshot and while it worked, 2.4 was still buggy and 5GHz didn't work.

We know it! The 5GHz radio is, at time, unsupported; and the 2.4GHz is in-dev. However, a lot of other devices are using the same radios. So we expect to have soon support for them (or you can try the closed-source drivers https://github.com/Nossiac/mtk-openwrt-feeds).

So, what we need now is: compile and test the last @mkresin 's staging tree:
https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary
If it works reliable, then a merge with "master" OpenWRT will be done.

Thank you all for your help! :smile:

mkresin commented:
@orangepizza 

> it look like leds doesn't have default setting.

Yepp, there is a typo. The double semicolon was missing at the end of the tplink,c2-v1 led block.

Would you please do a fresh clone from my [staging repository](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) and test again.

orangepizza commented:
wan and lan leds are both triggered by switch1, I think they should triggered by netdev/eth0.1 and netdev/eth0.2

dbeinder commented:
No, ethN.X are not guaranteed to be there and depend on the vlan(vId corresponds to X) configuration. the last parameter gives a bitmask for the physical switch ports.
That way, the LAN LED will light up if _any_ of the 4 LAN ports is active.
0x1e = 00011110  (bin)

orangepizza commented:
oh, luci setting of leds didn't show about bitmask, just the fact that its triggered to switch1 
![2018-06-23 07-19-51](https://user-images.githubusercontent.com/23009540/41801549-36d376ae-76b6-11e8-8ef8-0135048239ea.png)
luci bug?

mkresin commented:
> luci bug?

Yeah, it is either a LuCI bug or not implemented.

Anyway, thanks a lot for testing, I'll push my staging tree to master soon.

franzflasch commented:
Great! Thanks for all your work!

lars18th commented:
Hi @mkresin ,
Thank you for https://github.com/openwrt/openwrt/commit/8375623a064037bbdd914bd9e7529eefa7172560 !

Hi @franzflasch ,
Before to start to _improve_ this patch, now it's time to update the Wiki:
https://openwrt.org/toh/tp-link/archer_c2_ac750
Please, can someone upgrade these points?
- Update Support Status.
- Remove all LEDE references and point to OpenWRT Snapshots.
- Rewrite the description for the first time flash using the "c2_recovery.zip" file, for overcoming the need of compile/create a TFTP flash by hand.
- Include a link to this PR in the section Forum Threads for reference.

Regards.
